framework/src/modules/interoperability/sidechain/commands/initialize_state_recovery.ts

```diff
@@ -69,78 +70,90 @@ export class InitializeStateRecoveryCommand extends BaseInteroperabilityCommand<
 		validator.validate(chainDataSchema, deserializedSidechainAccount);
 
 		const mainchainAccount = await this.stores.get(ChainAccountStore).get(context, mainchainID);
-		// The commands fails if the sidechain is not terminated and did not violate the liveness requirement.
+		// The sidechain must either be terminated or must have violated the liveness limit while being active.
+		if (deserializedSidechainAccount.status === ChainStatus.REGISTERED) {
+			throw new Error('Sidechain has status registered.');
+		}
 		if (
-			deserializedSidechainAccount.status !== ChainStatus.TERMINATED &&
+			deserializedSidechainAccount.status === ChainStatus.ACTIVE &&
 			mainchainAccount.lastCertificate.timestamp -
 				deserializedSidechainAccount.lastCertificate.timestamp <=
 				LIVENESS_LIMIT
 		) {
-			throw new Error('Sidechain is not terminated.');
+			throw new Error('Sidechain is still active and obeys the liveness requirement.');
 		}
 
+		return {
+			status: VerifyStatus.OK,
+		};
+	}
+
+	// LIP: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0054.md#execution-3
+	public async execute(context: CommandExecuteContext<StateRecoveryInitParams>): Promise<void> {
+		const {
+			params: { chainID, siblingHashes, bitmap, sidechainAccount: sidechainAccountBuffer },
+		} = context;
+
+		const terminatedStateSubstore = this.stores.get(TerminatedStateStore);
+		const terminatedStateAccountExists = await terminatedStateSubstore.has(context, chainID);
+
 		const chainStore = this.stores.get(ChainAccountStore);
 		const queryKey = Buffer.concat([chainStore.key, utils.hash(chainID)]);
 
-		const query = { key: queryKey, value: utils.hash(sidechainAccount), bitmap };
+		const query = { key: queryKey, value: utils.hash(sidechainAccountBuffer), bitmap };
 
 		const proofOfInclusion = { siblingHashes, queries: [query] };
 
 		const smt = new SparseMerkleTree();
 		if (terminatedStateAccountExists) {
-			terminatedStateAccount = await terminatedStateSubstore.get(context, chainID);
-			if (!terminatedStateAccount.mainchainStateRoot) {
-				throw new Error('Sidechain account has missing property: mainchain state root');
-			}
+			const terminatedStateAccount = await terminatedStateSubstore.get(context, chainID);
 			const verified = await smt.verify(
 				terminatedStateAccount.mainchainStateRoot,
 				[queryKey],
 				proofOfInclusion,
 			);
 			if (!verified) {
+				this.events.get(InvalidSMTVerification).error(context);
```

Comment:
Personal taste, but I would refactor this a bit to avoid code repetition, for instance by creating a new variable `stateRoot` that can be equal to either `terminatedStateAccount.mainchainStateRoot` or `mainchainAccount.lastCertificate.stateRoot` and then write the verified check only once